Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCDav1dDecoder.cpp

```diff
@@ -0,0 +1,222 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "LibWebRTCDav1dDecoder.h"
+
+#if USE(LIBWEBRTC)
+
+#include "LibWebRTCMacros.h"
+#include "Logging.h"
+#include <algorithm>
+#include <wtf/FastMalloc.h>
+#include <wtf/UniqueRef.h>
+
+ALLOW_UNUSED_PARAMETERS_BEGIN
+ALLOW_COMMA_BEGIN
+
+#include <dav1d/dav1d.h>
+#include <libyuv/convert.h>
+#include <webrtc/api/scoped_refptr.h>
+#include <webrtc/api/video/encoded_image.h>
+#include <webrtc/api/video/i420_buffer.h>
+#include <webrtc/common_video/include/video_frame_buffer_pool.h>
+#include <webrtc/modules/video_coding/include/video_error_codes.h>
+#include <webrtc/rtc_base/logging.h>
+
+ALLOW_UNUSED_PARAMETERS_END
+ALLOW_COMMA_END
+
+namespace WebCore {
+
+class Dav1dDecoder final : public webrtc::VideoDecoder {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    Dav1dDecoder();
+    ~Dav1dDecoder();
+
+private:
+    bool Configure(const Settings&) final;
+    int32_t Decode(const webrtc::EncodedImage&, bool missingFrames, int64_t renderTimeMs) final;
+    int32_t RegisterDecodeCompleteCallback(webrtc::DecodedImageCallback*) final;
+    int32_t Release() final;
+    DecoderInfo GetDecoderInfo() const final;
+    const char* ImplementationName() const final;
+
+    webrtc::VideoFrameBufferPool m_bufferPool;
+    Dav1dContext* m_context { nullptr };
+    webrtc::DecodedImageCallback* m_decodeCompleteCallback { nullptr };
+};
+
+class ScopedDav1dData {
+public:
+    ~ScopedDav1dData() { dav1d_data_unref(&m_data); }
+    Dav1dData& Data() { return m_data; }
+
+private:
+    Dav1dData m_data;
+};
+
+class ScopedDav1dPicture {
+public:
+    ~ScopedDav1dPicture() { dav1d_picture_unref(&m_picture); }
+    Dav1dPicture& Picture() { return m_picture; }
+
+private:
+    Dav1dPicture m_picture;
+};
+
+constexpr char kDav1dName[] = "dav1d";
+
+// Calling `dav1d_data_wrap` requires a `free_callback` to be registered.
+static void NullFreeCallback(const uint8_t*, void*)
+{
+}
+
+Dav1dDecoder::Dav1dDecoder()
+    : m_bufferPool(false, 150)
+{
+}
+
+Dav1dDecoder::~Dav1dDecoder()
+{
+    Release();
+}
+
+bool Dav1dDecoder::Configure(const Settings& settings)
+{
+    Dav1dSettings s;
+    dav1d_default_settings(&s);
+
+    s.n_threads = std::max(2, settings.number_of_cores());
```

Comment:
This is all very confusing. We set multi-thread decoding , yet have a max_frame_delay of 1 ; so the entire decoding later heavily rely on this configuration.

If this was changed later, the entire code below would become unsafe, as dav1d could spawn another thread to perform the decoding on the compressed frame and queue it for later.

I think this deserves a comment so that it's not changed without understanding the consequences.

Reply:
Will add it. And we should indeed allow to use multithreading.